sdk/src/azure/core/az_mqtt5_rpc.c

```diff
@@ -67,234 +57,72 @@ AZ_NODISCARD AZ_INLINE bool _has_backslash_in_topic(az_span topic)
-        {
-          filter_remaining = az_span_slice_to_end(filter_remaining, az_span_size(filter_level));
-          filter_level = _get_next_topic_level(filter_remaining);
-          if (az_span_size(filter_level) == 1 && az_span_ptr(filter_level)[0] == '#')
-          {
-            return true;
-          }
-          return false;
-        }
-      }
-      else
-      {
-        return false;
-      }
-    }
-    else
-    {
-      if (!az_span_is_content_equal(filter_level, topic_level))
-      {
-        // Special case: Topic Filter = "foo/#" and Topic = "foo"
-        if (!_has_backslash_in_topic(topic_level) && _has_backslash_in_topic(filter_level))
-        {
-          az_span sub_no_backslash = az_span_slice(filter_level, 0, az_span_size(filter_level) - 1);
-          if (az_span_is_content_equal(sub_no_backslash, topic_level))
-          {
-            filter_remaining = az_span_slice_to_end(filter_remaining, az_span_size(filter_level));
-            filter_level = _get_next_topic_level(filter_remaining);
-            if (az_span_size(filter_level) == 1 && az_span_ptr(filter_level)[0] == '#')
-            {
-              return true;
-            }
-          }
-        }
-        return false;
-      }
-    }
-
-    filter_remaining = az_span_slice_to_end(filter_remaining, az_span_size(filter_level));
-    topic_remaining = az_span_slice_to_end(topic_remaining, az_span_size(topic_level));
-
-    filter_level = _get_next_topic_level(filter_remaining);
-    topic_level = _get_next_topic_level(topic_remaining);
-  }
-
-  if (az_span_size(topic_level) != 0)
-  {
-    return false;
-  }
-
-  return true;
-}
-
 AZ_NODISCARD bool az_mqtt5_rpc_status_failed(az_mqtt5_rpc_status status)
 {
   return (status < 200 || status >= 300);
 }
 
-AZ_NODISCARD static int32_t _az_rpc_calculate_substitution_length(
-    az_span format,
-    az_span key,
-    az_span value)
+AZ_INLINE AZ_NODISCARD bool _az_mqtt5_is_key_surrounded_by_slash(
+    int32_t idx,
+    int32_t key_size,
+    az_span topic_format)
 {
-  if (az_span_find(format, key) >= 0)
-  {
-    _az_PRECONDITION_VALID_SPAN(value, 1, false);
-    return az_span_size(value) - az_span_size(key);
-  }
-  return 0;
+  return (
+      (idx == 0 && (az_span_size(topic_format) == idx + key_size))
+      || (idx == 0 && az_span_ptr(topic_format)[idx + key_size] == '/')
+      || (idx > 0 && az_span_ptr(topic_format)[idx - 1] == '/'
+          && (az_span_size(topic_format) == idx + key_size))
+      || (idx > 0 && az_span_ptr(topic_format)[idx - 1] == '/'
+          && az_span_ptr(topic_format)[idx + key_size] == '/'));
 }
 
-AZ_NODISCARD static az_result _az_rpc_substitute_key_for_value(
-    az_span format,
-    uint8_t* format_buf,
-    az_span key,
-    az_span value,
-    az_span out_topic)
+AZ_NODISCARD bool _az_mqtt5_rpc_valid_topic_format(az_span topic_format)
 {
-  az_span temp_span = out_topic;
+  int32_t service_group_id_index = az_span_find(topic_format, service_group_id_key);
```

Comment:
Why not do each of these right before their if statement so if the function returns false on one of the first ones, we don't need to calculate the rest?

Reply:
Good point, changed it to calculate before if statements